Sources/_StringProcessing/Engine/MEBuiltins.swift

```diff
@@ -202,3 +115,159 @@ extension Processor {
     }
   }
 }
+
+// MARK: Built-in character class matching
+
+// Mentioned in ProgrammersManual.md, update docs if redesigned
+@_effects(releasenone)
+func _matchBuiltinCC(
+  _ cc: _CharacterClassModel.Representation,
+  in input: String,
+  at currentPosition: String.Index,
+  isInverted: Bool,
+  isStrictASCII: Bool,
+  isScalarSemantics: Bool
+) -> String.Index? {
+  guard currentPosition < input.endIndex else { 
+    return nil
+  }
+  if case .definite(let result) = _quickMatchBuiltinCC(
+    cc,
+    in: input,
+    at: currentPosition,
+    isInverted: isInverted,
+    isStrictASCII: isStrictASCII,
+    isScalarSemantics: isScalarSemantics
+  ) {
+    assert(result == _thoroughMatchBuiltinCC(
+      cc,
+      in: input,
+      at: currentPosition,
+      isInverted: isInverted,
+      isStrictASCII: isStrictASCII,
+      isScalarSemantics: isScalarSemantics))
+    return result
+  }
+  return _thoroughMatchBuiltinCC(
+    cc,
+    in: input,
+    at: currentPosition,
+    isInverted: isInverted,
+    isStrictASCII: isStrictASCII,
+    isScalarSemantics: isScalarSemantics)
```

Comment:
[Future] It would be great if we had a macro to expand into this pattern, since we have to essentially repeat the same function call three times. Something like:
```swift
return #quickMatch(
  _thoroughMatchBuiltinCC(
    cc,
    in: input,
    at: currentPosition,
    isInverted: isInverted,
    isStrictASCII: isStrictASCII,
    isScalarSemantics: isScalarSemantics))
```

Reply:
Yes, that would be nice. When I did the string re-gutting, I had some (always-inline) functions that took two different closures and applied this kind of pattern, but it turned too unwieldy and for the UTF-8 switch it ended up being better to just stamp the code out manually. A macro would be ideal here.